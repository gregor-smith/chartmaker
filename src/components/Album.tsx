import React, { FC } from 'react'
import { css, cx } from 'emotion'
import { Album as AlbumDetails, formatAlbumTitle } from '../state'


type Props = {
    details: AlbumDetails | null
    sizeRem: number
}


const baseStyle = css({
    backgroundColor: 'white',
    margin: '0.15rem'
})


export const Album: FC<Props> = ({ details, sizeRem }) => {
    const style = cx(
        baseStyle,
        css({
            width: `${sizeRem}rem`,
            height: `${sizeRem}rem`
        })
    )
    const image = details === null
        ? null
        : <img src={details.image.smallURL} alt={formatAlbumTitle(details)}/>
    return (
        <div className={style}>
            {image}
        </div>
    )
}
