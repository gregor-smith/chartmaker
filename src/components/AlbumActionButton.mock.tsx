const mock: typeof import('@/components/AlbumActionButton') = {
    AlbumActionButton: props =>
        <div {...props} className='mock-album-action-button'/>
}


export const AlbumActionButton = mock.AlbumActionButton
