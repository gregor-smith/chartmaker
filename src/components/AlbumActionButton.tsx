import { h, FunctionComponent } from 'preact'
import { css } from 'emotion'


type Props = {
    onClick: () => void
    title: string
}


const style = css({
    background: 'none',
    border: 'none'
})


export const AlbumActionButton: FunctionComponent<Props> = props =>
    <button {...props} class={style}/>
