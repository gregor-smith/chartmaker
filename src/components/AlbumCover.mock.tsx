import { cx } from 'emotion'


const mock: typeof import('@/components/AlbumCover') = {
    AlbumCover: ({
        album,
        size,
        overlayClass,
        children,
        highlighted,
        ...props
    }) => {
        const json = JSON.stringify(album)
        const className = cx('mock-album-cover', overlayClass)
        return (
            <div {...props} className={className}>
                {`Album: ${json}`}
                {`Size: ${size}`}
                {`Highlighted: ${highlighted}`}
                {children}
            </div>
        )
    }
}


export const AlbumCover = mock.AlbumCover
