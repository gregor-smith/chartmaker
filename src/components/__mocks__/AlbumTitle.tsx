import type { FC } from 'react'

import type { AlbumTitleProps } from '@/components/AlbumTitle'


export const AlbumTitle: FC<AlbumTitleProps> = ({ id, highlighted, children }) =>
    <div className='mock-album-title'>
        {`ID: ${id}`}
        {`Highlighted: ${highlighted}`}
        {children}
    </div>
