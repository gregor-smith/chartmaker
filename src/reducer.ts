import { SideEffectUpdate, Dispatch as _Dispatch } from './hooks'
import {
    State,
    createChart,
    SearchState,
    Album,
    escapeStateForExport,
    ChartShape,
    NamedAlbum
} from './state'
import {
    readFileText,
    findIndex,
    elementToDataURI,
    downloadURI,
    jsonToDataURI
} from './utils'
import { search } from './api'


type Action =
    | { tag: 'UpdateAPIKey', apiKey: string }
    | { tag: 'UpdateActiveChart', index: number }
    | { tag: 'PromptForNewChart' }
    | { tag: 'ShowChartNameTakenMessage' }
    | { tag: 'AddNewChart', name: string }
    | { tag: 'PromptToRenameActiveChart' }
    | { tag: 'RenameActiveChart', name: string }
    | { tag: 'PromptToDeleteActiveChart' }
    | { tag: 'DeleteActiveChart' }
    | { tag: 'ImportStateFile', file: File }
    | { tag: 'ShowInvalidStateImportMessage' }
    | { tag: 'LoadState', state: State }
    | { tag: 'PromptToExportState' }
    | { tag: 'CancelSearchRequest' }
    | { tag: 'SendSearchRequest' }
    | { tag: 'UpdateSearchState', state: SearchState }
    | { tag: 'UpdateSearchQuery', query: string }
    | { tag: 'DragChartAlbum', sourceID: number, targetID: number }
    | { tag: 'DropSearchAlbum', sourceID: number, targetID: number }
    | { tag: 'PromptToRenameAlbum', id: number }
    | { tag: 'RenameAlbum', id: number, name: string }
    | { tag: 'DeleteAlbum', id: number }
    | { tag: 'UpdateScreenshotLoading', loading: boolean }
    | { tag: 'UpdateScreenshotScale', scale: number }
    | { tag: 'TakeScreenshot', element: HTMLElement }
    | { tag: 'UpdateChartShape', shape: ChartShape, rowsX: number, rowsY: number }


export type ActionTag = Action['tag']
export type Dispatch<T extends ActionTag = ActionTag> = _Dispatch<Extract<Action, { tag: T }>>
export type DispatchProps<T extends ActionTag = ActionTag> = {
    dispatch: Dispatch<T>
}


export function reducer(state: State, action: Action): SideEffectUpdate<State, Action> {
    switch (action.tag) {
        case 'UpdateAPIKey':
            return {
                tag: 'Update',
                state: {
                    ...state,
                    apiKey: action.apiKey
                }
            }
        case 'UpdateActiveChart':
            return {
                tag: 'Update',
                state: {
                    ...state,
                    activeChartIndex: action.index
                }
            }
        case 'PromptForNewChart':
            return {
                tag: 'SideEffect',
                sideEffect: (dispatch, state) => {
                    const name = prompt('Enter new chart name:')?.trim()
                    if (name === undefined || name.length === 0) {
                        return
                    }
                    if (state.charts.some(chart => chart.name === name)) {
                        dispatch({ tag: 'ShowChartNameTakenMessage' })
                        return
                    }
                    dispatch({ tag: 'AddNewChart', name })
                }
            }
        case 'ShowChartNameTakenMessage':
            return {
                tag: 'SideEffect',
                sideEffect: () =>
                    alert('A chart with that name already exists')
            }
        case 'AddNewChart': {
            const [ albumIDCounter, chart ] = createChart({
                albumIDCounter: state.albumIDCounter,
                name: action.name
            })
            const charts = [ ...state.charts, chart ]
            return {
                tag: 'Update',
                state: {
                    ...state,
                    charts,
                    activeChartIndex: charts.length - 1,
                    albumIDCounter
                }
            }
        }
        case 'PromptToRenameActiveChart':
            return {
                tag: 'SideEffect',
                sideEffect: (dispatch, state) => {
                    const activeChart = state.charts[state.activeChartIndex]
                    const name = prompt(`Enter new name for '${activeChart.name}':`)?.trim()
                    if (name === undefined || name.length === 0) {
                        return
                    }
                    if (state.charts.some((chart, index) => index !== state.activeChartIndex && chart.name === name)) {
                        dispatch({ tag: 'ShowChartNameTakenMessage' })
                        return
                    }
                    dispatch({ tag: 'RenameActiveChart', name })
                }
            }
        case 'RenameActiveChart': {
            const activeChart = state.charts[state.activeChartIndex]
            return {
                tag: 'Update',
                state: {
                    ...state,
                    charts: [
                        ...state.charts.slice(0, state.activeChartIndex),
                        {
                            ...activeChart,
                            name: action.name
                        },
                        ...state.charts.slice(state.activeChartIndex + 1)
                    ]
                }
            }
        }
        case 'PromptToDeleteActiveChart':
            return {
                tag: 'SideEffect',
                sideEffect: dispatch => {
                    if (confirm('Really delete active chart? This cannot be undone')) {
                        dispatch({ tag: 'DeleteActiveChart' })
                    }
                }
            }
        case 'DeleteActiveChart': {
            if (state.charts.length === 1) {
                const [ albumIDCounter, chart ] = createChart({
                    albumIDCounter: state.albumIDCounter
                })
                return {
                    tag: 'Update',
                    state: {
                        ...state,
                        charts: [ chart ],
                        activeChartIndex: 0,
                        albumIDCounter
                    }
                }
            }

            const charts = [
                ...state.charts.slice(0, state.activeChartIndex),
                ...state.charts.slice(state.activeChartIndex + 1)
            ]

            return {
                tag: 'Update',
                state: {
                    ...state,
                    charts,
                    activeChartIndex: state.activeChartIndex - 1 < 0
                        ? charts.length - 1
                        : state.activeChartIndex  -1
                }
            }
        }
        case 'ImportStateFile':
            return {
                tag: 'SideEffect',
                sideEffect: async dispatch => {
                    try {
                        const json = await readFileText(action.file)
                        const parsed: unknown = JSON.parse(json)
                        const state = State.check(parsed)
                        dispatch({ tag: 'LoadState', state })
                    }
                    catch {
                        dispatch({ tag: 'ShowInvalidStateImportMessage' })
                    }
                }
            }
        case 'ShowInvalidStateImportMessage':
            return {
                tag: 'SideEffect',
                sideEffect: () =>
                    alert('Selected file is invalid')
            }
        case 'LoadState':
            return {
                tag: 'Update',
                state: action.state
            }
        case 'PromptToExportState':
            return {
                tag: 'SideEffect',
                sideEffect: (_dispatch, state) => {
                    const json = JSON.stringify(escapeStateForExport(state))
                    const uri = jsonToDataURI(json)
                    downloadURI(uri, 'state.json')
                }
            }
        case 'CancelSearchRequest': {
            if (state.search.tag !== 'Loading') {
                return { tag: 'NoUpdate' }
            }
            const controller = state.search.controller
            return {
                tag: 'UpdateWithSideEffect',
                state: {
                    ...state,
                    search: {
                        tag: 'Waiting',
                        query: state.search.query
                    }
                },
                sideEffect: () => controller.abort()
            }
        }
        case 'SendSearchRequest': {
            if (state.search.tag === 'Loading'
                    || state.search.query.trim().length === 0) {
                return { tag: 'NoUpdate' }
            }

            if (state.apiKey.trim().length === 0) {
                return {
                    tag: 'Update',
                    state: {
                        ...state,
                        search: {
                            tag: 'Error',
                            query: state.search.query,
                            message: 'Last.fm API key required'
                        }
                    }
                }
            }

            const controller = new AbortController()
            return {
                tag: 'UpdateWithSideEffect',
                state: {
                    ...state,
                    search: {
                        tag: 'Loading',
                        query: state.search.query,
                        controller
                    }
                },
                sideEffect: async (dispatch, state) => {
                    const result = await search({
                        key: state.apiKey,
                        query: state.search.query,
                        signal: controller.signal
                    })
                    switch (result.tag) {
                        case 'Ok': {
                            const albums: NamedAlbum[] = result.albums.map((album, index) => ({
                                ...album,
                                placeholder: false,
                                id: state.albumIDCounter + index + 1
                            }))
                            dispatch({
                                tag: 'LoadState',
                                state: {
                                    ...state,
                                    search: {
                                        tag: 'Complete',
                                        query: state.search.query,
                                        albums
                                    },
                                    albumIDCounter: state.albumIDCounter + result.albums.length
                                }
                            })
                            break
                        }
                        case 'StatusError': {
                            dispatch({
                                tag: 'UpdateSearchState',
                                state: {
                                    tag: 'Error',
                                    query: state.search.query,
                                    message: `Last.fm request returned ${result.status}`
                                }
                            })
                            break
                        }
                        case 'JSONDecodeError': {
                            dispatch({
                                tag: 'UpdateSearchState',
                                state: {
                                    tag: 'Error',
                                    query: state.search.query,
                                    message: 'Invalid data returned from Last.fm'
                                }
                            })
                            break
                        }
                        case 'NetworkError':
                            dispatch({
                                tag: 'UpdateSearchState',
                                state: {
                                    tag: 'Error',
                                    query: state.search.query,
                                    message: 'Network error sending request to Last.fm'
                                }
                            })
                        // 'Cancelled' is not handled because the only time the
                        // request is cancelled is when another is sent, in
                        // which case the search state is updated to 'Loading'
                        // anyway
                    }
                }
            }
        }
        case 'UpdateSearchState':
            return {
                tag: 'Update',
                state: {
                    ...state,
                    search: action.state
                }
            }
        case 'UpdateSearchQuery': {
            if (state.search.tag === 'Loading') {
                return { tag: 'NoUpdate' }
            }
            return {
                tag: 'Update',
                state: {
                    ...state,
                    search: {
                        ...state.search,
                        query: action.query
                    }
                }
            }
        }
        case 'DragChartAlbum': {
            if (action.sourceID === action.targetID) {
                return { tag: 'NoUpdate' }
            }

            const activeChart = state.charts[state.activeChartIndex]

            const sourceIndex = findIndex(activeChart.albums, album => album.id === action.sourceID)
            if (sourceIndex === null) {
                return { tag: 'NoUpdate' }
            }

            const targetIndex = findIndex(activeChart.albums, album => album.id === action.targetID)
            if (targetIndex === null) {
                return { tag: 'NoUpdate' }
            }

            let albums: Album[]
            if (sourceIndex < targetIndex) {
                albums = [
                    ...activeChart.albums.slice(0, sourceIndex),
                    ...activeChart.albums.slice(sourceIndex + 1, targetIndex + 1),
                    activeChart.albums[sourceIndex],
                    ...activeChart.albums.slice(targetIndex + 1)
                ]
            }
            else {
                albums = [
                    ...activeChart.albums.slice(0, targetIndex),
                    activeChart.albums[sourceIndex],
                    ...activeChart.albums.slice(targetIndex, sourceIndex),
                    ...activeChart.albums.slice(sourceIndex + 1)
                ]
            }

            return {
                tag: 'Update',
                state: {
                    ...state,
                    charts: [
                        ...state.charts.slice(0, state.activeChartIndex),
                        {
                            ...activeChart,
                            albums
                        },
                        ...state.charts.slice(state.activeChartIndex + 1)
                    ]
                }
            }
        }
        case 'DropSearchAlbum': {
            if (state.search.tag !== 'Complete') {
                return { tag: 'NoUpdate' }
            }

            const activeChart = state.charts[state.activeChartIndex]

            const source = state.search.albums.find(album => album.id === action.sourceID)
            if (source === undefined) {
                return { tag: 'NoUpdate' }
            }

            const targetIndex = findIndex(activeChart.albums, album => album.id === action.targetID)
            if (targetIndex === null) {
                return { tag: 'NoUpdate' }
            }

            return {
                tag: 'Update',
                state: {
                    ...state,
                    charts: [
                        ...state.charts.slice(0, state.activeChartIndex),
                        {
                            ...activeChart,
                            albums: [
                                ...activeChart.albums.slice(0, targetIndex),
                                {
                                    ...source,
                                    id: state.albumIDCounter + 1
                                },
                                ...activeChart.albums.slice(targetIndex + 1)
                            ]
                        },
                        ...state.charts.slice(state.activeChartIndex + 1)
                    ],
                    albumIDCounter: state.albumIDCounter + 1
                }
            }
        }
        case 'PromptToRenameAlbum':
            return {
                tag: 'SideEffect',
                sideEffect: (dispatch, state) => {
                    const album = state.charts[state.activeChartIndex]
                        .albums
                        .find(album => album.id === action.id)
                    if (album === undefined || album.placeholder) {
                        return
                    }

                    const name = prompt(`Enter new name for '${album.name}':`)?.trim()
                    if (name === undefined || name.length === 0) {
                        return
                    }
                    dispatch({
                        tag: 'RenameAlbum',
                        id: action.id,
                        name
                    })
                }
            }
        case 'RenameAlbum': {
            const activeChart = state.charts[state.activeChartIndex]

            const index = findIndex(activeChart.albums, album => album.id === action.id)
            if (index === null) {
                return { tag: 'NoUpdate' }
            }

            const album = activeChart.albums[index]
            if (album.placeholder) {
                return { tag: 'NoUpdate' }
            }

            return {
                tag: 'Update',
                state: {
                    ...state,
                    charts: [
                        ...state.charts.slice(0, state.activeChartIndex),
                        {
                            ...activeChart,
                            albums: [
                                ...activeChart.albums.slice(0, index),
                                {
                                    ...album,
                                    name: action.name
                                },
                                ...activeChart.albums.slice(index + 1)
                            ]
                        },
                        ...state.charts.slice(state.activeChartIndex + 1)
                    ]
                }
            }
        }
        case 'DeleteAlbum': {
            const activeChart = state.charts[state.activeChartIndex]

            const index = findIndex(activeChart.albums, album => album.id === action.id)
            if (index === null) {
                return { tag: 'NoUpdate' }
            }

            return {
                tag: 'Update',
                state: {
                    ...state,
                    charts: [
                        ...state.charts.slice(0, state.activeChartIndex),
                        {
                            ...activeChart,
                            albums: [
                                ...activeChart.albums.slice(0, index),
                                {
                                    id: action.id,
                                    placeholder: true
                                },
                                ...activeChart.albums.slice(index + 1)
                            ]
                        },
                        ...state.charts.slice(state.activeChartIndex + 1)
                    ]
                }
            }
        }
        case 'UpdateScreenshotLoading': {
            return {
                tag: 'Update',
                state: {
                    ...state,
                    screenshot: {
                        ...state.screenshot,
                        loading: action.loading
                    }
                }
            }
        }
        case 'UpdateScreenshotScale': {
            if (state.screenshot.loading) {
                return { tag: 'NoUpdate' }
            }
            return {
                tag: 'Update',
                state: {
                    ...state,
                    screenshot: {
                        ...state.screenshot,
                        scale: action.scale
                    }
                }
            }
        }
        case 'TakeScreenshot': {
            if (state.screenshot.loading) {
                return { tag: 'NoUpdate' }
            }
            return {
                tag: 'UpdateWithSideEffect',
                state: {
                    ...state,
                    screenshot: {
                        ...state.screenshot,
                        loading: true
                    }
                },
                sideEffect: async (dispatch, state) => {
                    const uri = await elementToDataURI(
                        action.element,
                        state.screenshot.scale
                    )
                    downloadURI(uri, 'chart.png')
                    dispatch({
                        tag: 'UpdateScreenshotLoading',
                        loading: false
                    })
                }
            }
        }
        case 'UpdateChartShape':
            return {
                tag: 'Update',
                state: {
                    ...state,
                    charts: [
                        ...state.charts.slice(0, state.activeChartIndex),
                        {
                            ...state.charts[state.activeChartIndex],
                            shape: action.shape,
                            rowsX: action.rowsX,
                            rowsY: action.rowsY
                        },
                        ...state.charts.slice(state.activeChartIndex + 1)
                    ]
                }
            }
    }
}
