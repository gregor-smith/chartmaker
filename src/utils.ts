import html2canvas from 'html2canvas'

import { BACKGROUND_COLOUR } from './style'


export function readFileText(file: File): Promise<string> {
    return new Promise((resolve, reject) => {
        const reader = new FileReader()
        reader.addEventListener('error', reject)
        reader.addEventListener('load', () => resolve(reader.result as string))
        reader.readAsText(file, 'utf-8')
    })
}


export function findIndex<T>(array: ReadonlyArray<T>, predicate: (item: T) => boolean): number | null {
    const index = array.findIndex(predicate)
    return index === -1 ? null : index
}


export async function elementToDataURI(element: HTMLElement, scale: number) {
    const canvas = await html2canvas(element, {
        allowTaint: true,
        scale,
        useCORS: true,
        imageTimeout: 0,
        backgroundColor: BACKGROUND_COLOUR
    })
    return canvas.toDataURL()
}


export function jsonToDataURI(json: string): string {
    return 'data:application/json;charset=utf-8,' + encodeURIComponent(json)
}


export function downloadURI(uri: string, filename: string) {
    const link = document.createElement('a')
    link.href = uri
    link.download = filename
    link.click()
    link.remove()
}
