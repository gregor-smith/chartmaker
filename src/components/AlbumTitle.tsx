import type { FC } from 'react'
import { css, cx } from 'emotion'

import type { DispatchProps } from '@/reducer'
import { RenameAlbumButton } from '@/components/RenameAlbumButton'
import { DeleteAlbumButton } from '@/components/DeleteAlbumButton'
import { SIDEBAR_LABEL_PADDING_SIZE, highlightBackgroundStyle } from '@/style'


const style = css({
    display: 'flex',
    [':not(:hover) button']: {
        display: 'none'
    }
})


const childrenStyle = css({
    marginRight: SIDEBAR_LABEL_PADDING_SIZE
})


export type AlbumTitleProps = DispatchProps & {
    id: number
    highlighted: boolean | undefined
}


export const AlbumTitle: FC<AlbumTitleProps> = ({
    dispatch,
    id,
    highlighted,
    children
}) => {
    function mouseEnter() {
        dispatch({
            tag: 'HighlightAlbum',
            targetID: id
        })
    }

    let className = style
    if (highlighted === false) {
        className = cx(style, highlightBackgroundStyle)
    }

    return (
        <div className={className} onMouseEnter={mouseEnter}>
            <span className={childrenStyle}>
                {children}
            </span>
            <RenameAlbumButton dispatch={dispatch} id={id}/>
            <DeleteAlbumButton dispatch={dispatch} id={id}/>
        </div>
    )
}
