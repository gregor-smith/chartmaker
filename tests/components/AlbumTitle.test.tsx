import React from 'react'
import { render } from 'react-dom'
import { act } from 'react-dom/test-utils'

import { AlbumTitle } from '@/components/AlbumTitle'

import { RenderContainer, ignore, fireEvent } from '../utils'
import { Action } from '@/reducer'


jest.mock('@/components/RenameAlbumButton')
jest.mock('@/components/DeleteAlbumButton')


const container = new RenderContainer()


test('renders title and two buttons', () => {
    render(
        <AlbumTitle dispatch={ignore} id={123} highlighted={undefined}>
            Test title
        </AlbumTitle>,
        container.element
    )

    expect(container.element).toMatchSnapshot()
})


test.each([ true, false ])('has reduced opacity when highlighted is false', highlighted => {
    render(
        <AlbumTitle dispatch={ignore} id={456} highlighted={highlighted}>
            Test title
        </AlbumTitle>,
        container.element
    )

    expect(container.element).toMatchSnapshot()
})


test('dispatches highlight action on mouse enter', () => {
    const mock = jest.fn<[], [ Action ]>()

    render(
        <AlbumTitle dispatch={mock} id={123} highlighted={undefined}>
            Test title
        </AlbumTitle>,
        container.element
    )

    act(() => fireEvent('mouseEnter', container.element?.firstChild))

    expect(mock).toHaveBeenCalledTimes(1)
    expect(mock).toHaveBeenCalledWith<[ Action ]>({
        tag: 'HighlightAlbum',
        targetID: 123
    })
})
