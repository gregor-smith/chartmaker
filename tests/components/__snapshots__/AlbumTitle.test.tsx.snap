// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`has reduced opacity when highlighted is false 1`] = `
.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-1:not(:hover) > *:not(:first-child) {
  display: none;
}

.emotion-0 {
  margin-right: 0.25rem;
}

<div
  id="test-render-container"
>
  <div
    class="emotion-1"
  >
    <span
      class="emotion-0"
    >
      Test name
    </span>
    Test children
  </div>
</div>
`;

exports[`has reduced opacity when highlighted is false 2`] = `
.emotion-0 {
  margin-right: 0.25rem;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  opacity: calc(100% / 3 * 2);
}

.emotion-1:not(:hover) > *:not(:first-child) {
  display: none;
}

<div
  id="test-render-container"
>
  <div
    class="emotion-1"
  >
    <span
      class="emotion-0"
    >
      Test name
    </span>
    Test children
  </div>
</div>
`;

exports[`renders title and children 1`] = `
.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-1:not(:hover) > *:not(:first-child) {
  display: none;
}

.emotion-0 {
  margin-right: 0.25rem;
}

<div
  id="test-render-container"
>
  <div
    class="emotion-1"
  >
    <span
      class="emotion-0"
    >
      Test name
    </span>
    Test children
  </div>
</div>
`;
